doc/sphinx/language/core/variants.rst

```diff
@@ -6,11 +6,30 @@ Variants and the `match` construct
 Variants
 --------
 
-.. cmd:: Variant @ident_decl {* @binder } {? %| {* @binder } } {? : @type } := {? %| } {+| @constructor } {? @decl_notations }
-
-   The :cmd:`Variant` command is similar to the :cmd:`Inductive` command, except
-   that it disallows recursive definition of types (for instance, lists cannot
-   be defined using :cmd:`Variant`). No induction scheme is generated for
+The :cmd:`Variant` command allows to define types by listing
+the inhabitants of the type. This includes enumerated types from
+programming languages, such as Booleans, characters or even the
+degenerate cases of the unit and empty types. Variant types more
+generally include enumerated types with arguments or even enumerated
+types with parametric arguments such as option types, sum types,
+etc. It also includes predicates or type families defined by cases
+such as the Boolean reflection or equality predicates. Observing the
+form of the inhabitants of a variant type is done by case analysis
+using the :cmd:`match` expression. Each case is described by what is
+called a :gdef:`constructor` of the type. For instance, Booleans have two
+constructors: :g:`true` and :g:`false`.
+
+When constructors recursively take arguments in the type being
+defined, the type becomes recursive, in which case it can be either
+:cmd:`Inductive` or :cmd:`CoInductive`. The terminology :cmd:`Variant`
+is thus specifically used to indicate a non-recursive type defined by
+constructors (for instance, natural numbers, lists or streams cannot
```

Comment:
"stream" (an infinite sequence of numbers?) may not be familiar to your average software engineer.  I didn't know it.

Reply:
In Coq, streams are the following type:
```coq
CoInductive Stream (A : Type) := { hd : A ; tl : Stream A }.
```
This indeed generalizes infinite sequence of numbers to infinite sequence of any kind of objects.

Then, [coinductive types](https://en.wikipedia.org/wiki/Coinduction#Codata) is a generalization of streams.